hal/armv7r/pmap.c

```diff
@@ -0,0 +1,288 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * pmap - machine dependent part of VM subsystem (ARMv7r)
+ *
+ * Copyright 2017, 2020-2022, 2024 Phoenix Systems
+ * Author: Pawel Pisarczyk, Aleksander Kaminski, Hubert Buczynski, Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "hal/pmap.h"
+
+#include "config.h"
+#include "syspage.h"
+#include "halsyspage.h"
+#include <arch/cpu.h>
+#include <arch/spinlock.h>
+
+
+/* Linker symbols */
+extern unsigned int _end;
+extern unsigned int __bss_start;
+
+u8 _init_stack[NUM_CPUS][SIZE_INITIAL_KSTACK] __attribute__((aligned(8)));
+
+
+static struct {
+	unsigned int kernelCodeRegion;
+	spinlock_t lock;
+} pmap_common;
+
+
+static void pmap_mpu_setMemRegionNumber(u32 num)
+{
+	__asm__ volatile("mcr p15, 0, %0, c6, c2, 0" ::"r"(num));
+}
+
+
+static void pmap_mpu_setMemRegionRasr(u32 rasr)
+{
+	/* ARMv7-R uses the same region attribute bits as ARMv7-M, but they are split over two registers */
+	u32 rser = rasr & 0xffff;
+	u32 racr = rasr >> 16;
+	__asm__ volatile("mcr p15, 0, %0, c6, c1, 2" ::"r"(rser));
+	__asm__ volatile("mcr p15, 0, %0, c6, c1, 4" ::"r"(racr));
+}
+
+
+static void pmap_mpu_setMemRegionStatus(int enable)
+{
+	u32 val;
+	__asm__ volatile("mrc p15, 0, %0, c6, c1, 2" :"=r"(val));
+	if (enable != 0) {
+		val |= 1;
+	}
+	else {
+		val &= ~1;
+	}
+
+	__asm__ volatile("mcr p15, 0, %0, c6, c1, 2" ::"r"(val));
+}
+
+
+static void pmap_mpu_setMemRegionRbar(u32 addr)
+{
+	addr &= ~((1 << 5) - 1);
+	asm volatile("mcr p15, 0, %0, c6, c1, 0" ::"r"(addr));
```

Comment:
Let's stick to a single style (I'm ignoring `volatile(`, clang-format is very persistent in its error)
```suggestion
	__asm__ volatile("mcr p15, 0, %0, c6, c1, 0" ::"r"(addr));
```